SSZ-QL: update "path parsing" data types

fernantho opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
This PR improves data types for handling paths in SSZ-QL queries and clarifies how list lengths are parsed.
- **New `Path` type**:
```golang
type Path struct {
	Length   bool
	Elements []PathElement
}
```
- **Updated the `PathElement` type** (removes the `Length`):
```golang
type PathElement struct {
	Name string
	// [Optional] Index for List/Vector elements
	Index *uint64
}
```
With this change:
- A “length query” is always terminal, an SSZ-QL path cannot contain more than one len request.
- Path processing is simpler and more efficient, as the queries fit into types

It has also been changed how lengths are requested. In the prior version, to query the length of a `List` or `Bitlist`, we followed the following convention:
```
block.payload.len(transactions)
```
It now has been changed to the following one:
```
len(block.payload.transactions)
```
where length is tied to the list `PathElement` item into the `Path`, in the prior example, the length is tied to `transactions` field.

**Which issues(s) does this PR fix?**
It is related to https://github.com/OffchainLabs/prysm/issues/15598 and https://github.com/OffchainLabs/prysm/issues/15587:

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

fernantho committed:
updated path processing data types, refactored ParsePath and fixed tests

fernantho committed:
updated generalized index accordingly, changed input parameter path type from []PathElemen to Path

fernantho committed:
updated query.go accordingly, changed input parameter path type from []PathElemen to Path

fernantho committed:
added descriptive changelog

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -10,14 +10,20 @@ import (
 	Index *uint64
 }
 
+type Path struct {
```
Would you mind adding a documentation for this newly introduced struct?

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -26,8 +32,24 @@ var lengthRegex = regexp.MustCompile(`^\s*len\s*\(\s*([^)]+?)\s*\)\s*$`)
+	// 2. Inner expression: "field_name"
+	if len(matches) == 2 {
+		processedPath.Length = true
+		// If we have o¡found a len() expression, we only want to parse its inner expression.
```
```suggestion
		// If we have found a len() expression, we only want to parse its inner expression.
```

syjn99 commented on encoding/ssz/query/path.go:
```diff
@@ -37,24 +59,13 @@ func ParsePath(rawPath string) ([]PathElement, error) {
 	var path []PathElement
```
```suggestion
	var elem []PathElement
```
`path` seems quite confusing at this PR. I'm not so sure what's best option: `pathElements`, `elements`, etc.

fernantho replied:
Totally agree with this. Renaming this to `pathElements` because it is the most explicit

syjn99 reviewed (commented):
Looks great!

fernantho committed:
Update encoding/ssz/query/path.go

Co-authored-by: Jun Song <87601811+syjn99@users.noreply.github.com>

fernantho committed:
Added documentation for Path struct and renamed  to  for clarity

fernantho committed:
Merge branch 'develop' into fix/refactor-path-parsing

rkapka commented on changelog/fernantho_ssz-ql-update-path-processing.md:
```diff
@@ -0,0 +1,3 @@
+## Added
```
nitpick: I think this PR belongs to the `Changed` category

rkapka commented on encoding/ssz/query/generalized_index.go:
```diff
@@ -53,7 +53,7 @@ func GetGeneralizedIndexFromPath(info *SszInfo, path []PathElement) (uint64, err
 		currentInfo = fieldSsz
 
 		// Check if a path element is a length field
```
This comment is outdated

rkapka commented on encoding/ssz/query/path.go:
```diff
@@ -26,35 +36,40 @@ var lengthRegex = regexp.MustCompile(`^\s*len\s*\(\s*([^)]+?)\s*\)\s*$`)
 	for _, elem := range rawElements {
 		if elem == "" {
-			return nil, errors.New("invalid path: consecutive dots or trailing dot")
+			return Path{}, errors.New("invalid pathElements: consecutive dots or trailing dot")
```
```suggestion
			return Path{}, errors.New("invalid path: consecutive dots or trailing dot")
```

fernantho committed:
Update encoding/ssz/query/path.go

Co-authored-by: Radosław Kapka <radoslaw.kapka@gmail.com>

fernantho committed:
updated changelog to its correct type: Changed

fernantho committed:
updated outdated comment in generalized_index.go and removed test in generalized_index_test.go as this one belongs in path_test.go

fernantho committed:
Added validateRawPath with strict raw-path validation only - no raw-path fixing is added. Added test suite covering

fernantho committed:
added extra tests for wrongly formated paths

fernantho commented:
Following the discussion with @rkapka:
<img width="938" height="116" alt="image" src="https://github.com/user-attachments/assets/1a853e5c-8a2e-4fd6-8df6-edf15c8fb3ad" />
I added a `validateRawPath` function that performs strict format checks on the raw path string and updated the tests.

**What validateRawPath enforces**:
1. Rejects any invalid character (allowed: letters, digits, ., [], (), _; whitespace is rejected).
2. Parentheses are balanced (count-based).
3. Brackets are balanced (count-based).
4. len( … ) is only allowed at the very start of the path.
5. Empty len() is rejected.
6. Malformed bracket transitions (e.g., "][a", "][_") are rejected.

```
data P
.data P
data. F
.data. F
..data F
data.target.root P
len(data) P
len(.data) P
len(data.) F
.len(data) F
len(data.target.root) P
data.target.len(root) F
data.len P (len can be a valid identifier)
data.len.value P
len.len P
len.len(root) F
len() F
len(data).foo F
len(data F
len(root)) F
len(a,b) F
arr[42] P
arr[18446744073709551615] P
arr[42]foo F
arr[42].foo P
arr[42].foo.bar[10] P
len(arr[42]) P
len(arr)[0] F
len( arr[42].foo ) P
data . target F
. data F
len(. data) F
```

**Notes**
- The function validates format only; it does not try to “fix” input.
- We intentionally do not strip whitespace or auto-correct the raw string; any non-valid input is rejected.

fernantho committed:
Merge branch 'develop' into fix/refactor-path-parsing

rkapka committed:
Merge branch 'develop' into fix/refactor-path-parsing
